File: backend/models/content/news.py
import uuid
from typing import List
from sqlalchemy import Column, ForeignKey, Integer, String
from sqlalchemy.dialects.postgresql import ARRAY, UUID
from sqlalchemy import inspect
from utility.database import db
from utility.constants import AVAILABLE_LANGUAGES
from utility.translation import get_translation
from models.content.base import Item


class News(Item):
    """
    News model which inherits from the base Item model.

    Attributes:
        news_id: Primary key
    """

    news_id = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    url = Column(String(length=512), nullable=True, index=True)

    # Foreign keys
    item_id = Column(UUID(as_uuid=True), ForeignKey("item.item_id"))

    # Relationships
    item = db.relationship("Item", back_populates="news")
    translations = db.relationship("NewsTranslation", back_populates="news")

    __mapper_args__ = {"polymorphic_identity": "news"}

    def to_dict(
        self, provided_languages: List[str] = AVAILABLE_LANGUAGES, is_public_route=True
    ):
        base_data = super().to_dict(
            provided_languages=provided_languages, is_public_route=is_public_route
        )

        if not base_data:
            return None

        translations = []

        for language_code in provided_languages:
            translation = get_translation(
                NewsTranslation, ["news_id"], {"news_id": self.news_id}, language_code
            )
            translations.append(translation)

        del base_data["news_id"]

        base_data["translations"] = [
            translation.to_dict() for translation in translations
        ]

        return base_data


class NewsTranslation(db.Model):
    __tablename__ = "news_translation"

    news_translation_id = Column(
        UUID(as_uuid=True), primary_key=True, default=uuid.uuid4
    )

    title = Column(String(255))
    body = Column(String(100_000))
    short_description = Column(String(255))
    main_image_url = Column(String(2096))
    sub_image_urls = Column(ARRAY(String))

    # Foreign keys
    news_id = Column(UUID(as_uuid=True), ForeignKey("news.news_id"))
    language_code = Column(String(20), ForeignKey("language.language_code"))

    # Relationships
    news = db.relationship("News", back_populates="translations")
    language = db.relationship("Language", back_populates="news_translations")

    def to_dict(self):
        columns = inspect(self.__class__)

        if not columns:
            return {}

        columns = columns.mapper.column_attrs.keys()

        data = {}
        for column in columns:
            data[column] = getattr(self, column)

        del data["news_translation_id"]
        del data["news_id"]

        return data

File: frontend/src/models/items/News.ts
import Item from '@/models/Items'
import { LanguageCode } from '@/models/Language'

/**
 * @interface News
 * @extends Item
 * @description Describes a news item from the backend and API responses
 *
 * @property {string} url - The URL of the news item
 * @property {NewsTranslation[]} translations - The translations of the news item
 */
export default interface News extends Item {
  url: string
  translations: NewsTranslation[]
}

/**
 * @interface NewsTranslation
 * @description Holds the translations of each news item, is at least 1:1 with the News model and is used to display the news item in different languages
 *
 * @property {string} title - The title of the news item
 * @property {string} body - The body of the news item
 * @property {LanguageCode} language_code - The language code of the news item
 * @property {string} main_image_url - The main image URL of the news item
 * @property {string} short_description - The short description of the news item
 * @property {string[]} sub_image_urls - The sub image URLs of the news item (optional)
 */
export interface NewsTranslation {
  title: string
  body: string
  language_code: LanguageCode
  main_image_url: string
  short_description: string
  sub_image_urls?: string[]
}
